palace/utils/communication.hpp

```diff
@@ -219,6 +245,24 @@ class Mpi
     GlobalOp(len, buff, MPI_SUM, comm);
   }
 
+  // Return the rank that has the minimum over the communicator
+  template <typename T>
+  static mpi::ValueAndRank<T> FindMin(const T val, MPI_Comm comm)
+  {
+    mpi::ValueAndRank<T> local{val, Mpi::Rank(comm)};
```

Comment:
I'd propose the following changes to these wrappers to align better with the other ones and the MPI API:

```
diff --git a/palace/utils/communication.hpp b/palace/utils/communication.hpp
index 09febd99..a7df5855 100644
--- a/palace/utils/communication.hpp
+++ b/palace/utils/communication.hpp
@@ -5,6 +5,7 @@
 #define PALACE_UTILS_COMMUNICATION_HPP

 #include <complex>
+#include <vector>
 #include <fmt/format.h>
 #include <fmt/printf.h>
 #include <fmt/ranges.h>
@@ -127,31 +128,49 @@ inline MPI_Datatype DataType<bool>()
   return MPI_C_BOOL;
 }

-// Helper for returning the rank with a MIN or MAX reduction.
-template <typename T>
-struct ValueAndRank
+template <typename T, typename U>
+struct ValueAndLoc
 {
-  // The reduced value from the MINLOC or MAXLOC operation.
-  T value;
-  // The processor rank that has the value.
-  int rank;
-  // Convenience decay operators to allow use as the underlying value.
-  inline operator const T &() const { return value; }
-  inline operator T &() { return value; }
+  T val;
+  U loc;
 };

 template <>
-inline MPI_Datatype DataType<ValueAndRank<double>>()
+inline MPI_Datatype DataType<ValueAndLoc<float, signed int>>()
+{
+  return MPI_FLOAT_INT;
+}
+
+template <>
+inline MPI_Datatype DataType<ValueAndLoc<double, signed int>>()
 {
   return MPI_DOUBLE_INT;
 }

 template <>
-inline MPI_Datatype DataType<ValueAndRank<int>>()
+inline MPI_Datatype DataType<ValueAndLoc<long double, signed int>>()
+{
+  return MPI_LONG_DOUBLE_INT;
+}
+
+template <>
+inline MPI_Datatype DataType<ValueAndLoc<signed short, signed int>>()
+{
+  return MPI_SHORT_INT;
+}
+
+template <>
+inline MPI_Datatype DataType<ValueAndLoc<signed int, signed int>>()
 {
   return MPI_2INT;
 }

+template <>
+inline MPI_Datatype DataType<ValueAndLoc<signed long int, signed int>>()
+{
+  return MPI_LONG_INT;
+}
+
 }  // namespace mpi

 //
@@ -169,7 +188,6 @@ public:
   // Finalize MPI (if it has been initialized and not yet already finalized).
   static void Finalize()
   {
-
     if (IsInitialized() && !IsFinalized())
     {
       MPI_Finalize();
@@ -245,22 +263,40 @@ public:
     GlobalOp(len, buff, MPI_SUM, comm);
   }

-  // Return the rank that has the minimum over the communicator
-  template <typename T>
-  static mpi::ValueAndRank<T> FindMin(const T val, MPI_Comm comm)
+  // Global minimum with location (in-place, result is broadcast to all processes).
+  template <typename T, typename U>
+  static void GlobalMinLoc(int len, T *val, U *loc, MPI_Comm comm)
   {
-    mpi::ValueAndRank<T> local{val, Mpi::Rank(comm)};
-    GlobalOp(1, &local, MPI_MINLOC, comm);
-    return local;
+    std::vector<mpi::ValueAndLoc<T, U>> buff(len);
+    for (int i = 0; i < len; i++)
+    {
+      buff[i].val = val[i];
+      buff[i].loc = loc[i];
+    }
+    GlobalOp(len, buff.data(), MPI_MINLOC, comm);
+    for (int i = 0; i < len; i++)
+    {
+      val[i] = buff[i].val;
+      loc[i] = buff[i].loc;
+    }
   }

-  // Return the rank that has the maximum over the communicator
+  // Global maximum with location (in-place, result is broadcast to all processes).
   template <typename T>
-  static mpi::ValueAndRank<T> FindMax(const T &val, MPI_Comm comm)
+  static void GlobalMaxLoc(int len, T *val, U *loc, MPI_Comm comm)
   {
-    mpi::ValueAndRank<T> local{val, Mpi::Rank(comm)};
-    GlobalOp(1, &local, MPI_MAXLOC, comm);
-    return local;
+    std::vector<mpi::ValueAndLoc<T, U>> buff(len);
+    for (int i = 0; i < len; i++)
+    {
+      buff[i].val = val[i];
+      buff[i].loc = loc[i];
+    }
+    GlobalOp(len, buff.data(), MPI_MAXLOC, comm);
+    for (int i = 0; i < len; i++)
+    {
+      val[i] = buff[i].val;
+      loc[i] = buff[i].loc;
+    }
   }

   // Global broadcast from root.
```